src/composer/prover.rs

```diff
@@ -0,0 +1,511 @@
+        bytes = &bytes[8..];
+
+        let prover_key_len =
+            <[u8; 8]>::try_from(&bytes[..8]).expect("checked len");
+        let prover_key_len = u64::from_be_bytes(prover_key_len) as usize;
+        bytes = &bytes[8..];
+
+        let commit_key_len =
+            <[u8; 8]>::try_from(&bytes[..8]).expect("checked len");
+        let commit_key_len = u64::from_be_bytes(commit_key_len) as usize;
+        bytes = &bytes[8..];
+
+        let verifier_key_len =
+            <[u8; 8]>::try_from(&bytes[..8]).expect("checked len");
+        let verifier_key_len = u64::from_be_bytes(verifier_key_len) as usize;
+        bytes = &bytes[8..];
+
+        let size = <[u8; 8]>::try_from(&bytes[..8]).expect("checked len");
+        let size = u64::from_be_bytes(size) as usize;
+        bytes = &bytes[8..];
+
+        let constraints =
+            <[u8; 8]>::try_from(&bytes[..8]).expect("checked len");
+        let constraints = u64::from_be_bytes(constraints) as usize;
+        bytes = &bytes[8..];
+
+        if bytes.len()
+            < label_len + prover_key_len + commit_key_len + verifier_key_len
+        {
+            return Err(Error::NotEnoughBytes);
+        }
+
+        let label = &bytes[..label_len];
+        bytes = &bytes[label_len..];
+
+        let prover_key = &bytes[..prover_key_len];
+        bytes = &bytes[prover_key_len..];
+
+        let commit_key = &bytes[..commit_key_len];
+        bytes = &bytes[commit_key_len..];
+
+        let verifier_key = &bytes[..verifier_key_len];
+
+        let label = label.to_vec();
+        let prover_key = ProverKey::from_slice(prover_key)?;
+
+        // Safety: checked len
+        let commit_key =
+            unsafe { CommitKey::from_slice_unchecked(&commit_key) };
+
+        let verifier_key = VerifierKey::from_slice(verifier_key)?;
+
+        Ok(Self::new(
+            label,
+            prover_key,
+            commit_key,
+            verifier_key,
+            size,
+            constraints,
+        ))
+    }
+
+    /// Prove the circuit
+    pub fn prove<R>(
+        &self,
+        rng: &mut R,
+        circuit: &C,
+    ) -> Result<(Proof, Vec<BlsScalar>), Error>
+    where
+        C: Circuit,
+        R: RngCore + CryptoRng,
+    {
+        let prover = Builder::prove(self.constraints, circuit)?;
+
+        let constraints = self.constraints;
+        let size = self.size;
+
+        let domain = EvaluationDomain::new(constraints)?;
+
+        let mut transcript = self.transcript.clone();
+
+        let public_inputs = prover.public_inputs();
+        let public_input_indexes = prover.public_input_indexes();
+        let dense_public_inputs = Builder::dense_public_inputs(
+            &public_input_indexes,
+            &public_inputs,
+            self.size,
+        );
+
+        public_inputs
+            .iter()
+            .for_each(|pi| transcript.append_scalar(b"pi", pi));
+
+        // round 1
+        // convert wires to padded scalars
```

Comment:
I think the original comment

> // Note that `d_w` is added for the additional selector for 3-input
> // gates `q_4`.

is still worth mentioning here

Reply:
Hmm you think so? Our implementation doesn't make the fourth wire optional. Its not an implementation detail for us but the basis of our protocol. We have other special gates that aren't mentioned such as the logic, range, variable/fixed group.